docs/software-architecture-and-design.md

```diff
@@ -289,6 +294,56 @@ To summarize, the best practices are as follows:
 2. Avoid using coroutines with resizeable containers. If they must be used, make sure that the coroutines are allocated on the heap.
 3. Pass data to the coroutine on creation as much as possible, avoid using member variables.
 
+# Finite State Machines
+## What Are Finite State Machines?
+A finite state machine (FSM) is a system with a finite number of states with defined transitions and outputs based on the inputs to the system. In particular, we are interested in hierarchical state machines where we can transitions between states in terms of _when_ states should transition (_guards_) and _what_ should happen when transitions occur (_actions_), given a specific input (_event_). Hierarchical state machines treat a sub-FSM as a state with _guards_ and _actions_ when transitioning to and from the sub-FSM. When the sub-FSM enters a terminal state, the parent FSM is able to automatically transition to another state.
```

Comment:
typo: change to "transition" instead of "transitions" in the sentence "In particular, ... where we can transitions between..."

Reply:
fixed